### PR TITLE
Use `Compilation_unit.t` instead of `Ident.t` for globals

### PR DESCRIPTION
**Pull request by lukemaurer**
This PR adds a requirement to use `-for-pack` when compiling an .mli. In return, we will now always be able to determine the fully-qualified name of a compilation unit during compilation, rather than having to fix up identifiers when assembling a pack. Because of this, it makes more sense to use `Compilation_unit.t` instead of `Ident.t` whenever an identifier is known to be global. Currently this is done in the type checker and in Lambda - in particular, the `Pgetglobal` and `Psetglobal` primitives now operate on `Compilation_unit.t` rather than `Ident.t`. (There is a new primitive, `Pgetpredef`, for getting predefined values, since these have no representation as `Compilation_unit.t`s.) This simplifies a number of things, not least `asmpackager.ml` and `bytepackager.ml`, which no longer have to do any rewriting (and `export_info_for_pack.ml` is gone entirely).

### CODE REVIEW

**Comment by lukemaurer** [comment]
Coverting to draft, since #896 is a subset of this, and I'll probably pull more pieces off of it.

File: ocaml/typing/env.ml [hunk 1]
```diff
@@ -1183,9 +1214,14 @@ and expand_module_path lax env path =
...
       if lax || !Clflags.transparent_modules then path' else
-      let id = Path.head path in
-      if Ident.is_global_or_predef id && not (Ident.same id (Path.head path'))
-      then add_required_global id;
+      let head = address_head (find_value_address path env) in
+      let head' = address_head (find_value_address path' env) in
+      begin match head with
+      | AHunit cu when not (equal_address_head head head') ->
+        add_required_global cu
+      | AHunit _ | AHlocal _ ->
```
Comment (lpw25) [inline ocaml/typing/env.ml:1222]: I think that this is adding the wrong thing to the required global. The point of this code is to make sure that a reference to `Foo.Bar` will force `Foo` to get linked -- even if `Bar` is a transparent alias to a module in another file. Whereas this new version asks for the address of `Foo.Bar` which will not be `Foo` if it's a transparent alias.

File: ocaml/typing/env.ml [hunk 2]
```diff
@@ -1524,16 +1560,18 @@ let add_to_tbl id decl tbl =
   let decls = try NameMap.find id tbl with Not_found -> [] in
   NameMap.add id (decl :: decls) tbl
 
+exception Primitives_don't_have_addresses
+
 let value_declaration_address (_ : t) id decl =
   match decl.val_kind with
-  | Val_prim _ -> EnvLazy.create_failed Not_found
-  | _ -> EnvLazy.create_forced (Aident id)
+  | Val_prim _ -> EnvLazy.create_failed Primitives_don't_have_addresses
```
Comment (lpw25) [inline ocaml/typing/env.ml:1567]: Why the change in exception?

File: ocaml/typing/env.ml [hunk 3]
```diff
@@ -1134,6 +1158,14 @@ let find_constructor_address path env =
   | Papply _ ->
       raise Not_found
 
+let find_compilation_unit modname =
+  (* CR-someday lmaurer: This is a bit gross. Could add an
+     [mda_compilation_unit] field, but it's not clear how
+     to fill it in other cases. *)
+  match get_address (find_pers_mod modname).mda_address with
+  | Aunit cu -> cu
+  | Alocal _ | Adot _ -> assert false (* see [sign_of_cmi] *)
```
Comment (lpw25) [inline ocaml/typing/env.ml:1167]: Why is this function needed? Why not add `find_compilation_unit` to `Persistent_env` and then call that?

File: ocaml/typing/printtyp.mli [hunk 4]
```diff
@@ -181,6 +181,8 @@ val print_items: (Env.t -> signature_item -> 'a option) ->
    for Foo__bar. This pattern is used by the stdlib. *)
 val rewrite_double_underscore_paths: Env.t -> Path.t -> Path.t
 
+val rewrite_double_underscore_longidents: Env.t -> Longident.t -> Longident.t
+
```
Comment (lpw25) [inline ocaml/typing/printtyp.mli:185]: Why is this function needed?

**Review by lpw25 — COMMENTED** [review]
I've reviewed the typing changes. Mostly fine. I think that the `required_globals` isn't quite right and I have various small comments.

File: ocaml/typing/types.ml [hunk 5]
```diff
@@ -100,8 +100,8 @@ end
 
 module Uid = struct
   type t =
-    | Compilation_unit of string
-    | Item of { comp_unit: string; id: int }
+    | Compilation_unit of Compilation_unit.t
+    | Item of { comp_unit: Compilation_unit.t option; id: int }
```
[the remark this answers is not in the document]
Comment (lukemaurer) [reply ocaml/typing/types.ml:104]: Changed. I assume it's the runtime representation that's the issue for Merlin, not the construction API?

File: ocaml/typing/typemod.ml [hunk 6]
```diff
@@ -1976,9 +1976,17 @@ and type_module_aux ~alias sttn funct_body anchor env smod =
...
+      if alias && aliasable then begin
+        if not !Clflags.transparent_modules then begin
+          (* -no-alias-deps is off; we can count on having the .cmi available to
+             get the full compilation unit *)
+          let address = Env.find_module_address path env in
+          match Env.address_head address with
+          | AHunit cu -> Env.add_required_global cu
+          | AHlocal _ -> ()
+        end;
+        md
```
[the remark this answers is not in the document]
Comment (lukemaurer) [reply ocaml/typing/typemod.ml:1988]: I've made `Env.add_required_global` take a `Path.t`, leaving this code mostly where it was before.

File: ocaml/typing/typemod.ml [hunk 7]
```diff
@@ -2887,13 +2897,19 @@ let package_units initial_env objfiles cmifile modulename =
     List.map
       (fun f ->
          let pref = chop_extensions f in
-         let modname = String.capitalize_ascii(Filename.basename pref) in
+         let modname =
+           String.capitalize_ascii(Filename.basename pref)
+           |> Compilation_unit.Name.of_string
+           |> Compilation_unit.create_child modulename
```
[the remark this answers is not in the document]
Comment (lukemaurer) [reply ocaml/typing/typemod.ml:2903]: Changed

File: ocaml/typing/env.ml (hunk 1 above)
Comment (lukemaurer) [reply ocaml/typing/env.ml:1222]: Now that `add_required_global` takes a `Path.t`, I've put this back to essentially its original form.

File: ocaml/typing/env.ml [hunk 8]
```diff
@@ -735,7 +751,13 @@ let find_same_module id tbl =
 let find_name_module ~mark name tbl =
   match IdTbl.find_name wrap_module ~mark name tbl with
   | x -> x
-  | exception Not_found when not (Current_unit_name.is name) ->
+  | exception Not_found
+    when not (Current_unit_name.is (name |> Compilation_unit.Name.of_string)) ->
```
[the remark this answers is not in the document]
Comment (lukemaurer) [reply ocaml/typing/env.ml:755]: Done.

File: ocaml/typing/env.ml (hunk 2 above)
Comment (lukemaurer) [reply ocaml/typing/env.ml:1567]: I should have written it down, but as I recall, there was something that was tricky to debug because the `Not_found` was caught. I suppose `Fatal_error "Primitives don't have addresses"` would be better, since we don't intend that anyone catch the exception.

File: ocaml/typing/env.ml (hunk 3 above)
Comment (lukemaurer) [reply ocaml/typing/env.ml:1167]: Looks like it isn't needed after all, actually (thankfully). Getting rid of it.

File: ocaml/typing/printtyp.mli (hunk 4 above)
Comment (lukemaurer) [reply ocaml/typing/printtyp.mli:185]: Because `transl_extension_constructor` (among many other things) now takes a `Longident.t` rather than a `Path.t` and it needs to do the same rewrite it did before.

File: ocaml/typing/env.ml (hunk 2 above)
Comment (lukemaurer) [reply ocaml/typing/env.ml:1567]: Changed to `Invalid_argument "Primitives don't have addresses"`.

File: ocaml/typing/persistent_env.ml [hunk 9]
```diff
@@ -193,6 +197,18 @@ let acknowledge_pers_struct penv check modname pers_sig pm =
...
+  begin match Compilation_unit.get_current () with
+  | Some current_unit ->
+      let access_allowed =
+        Compilation_unit.can_access_by_name name ~accessed_by:current_unit
+      in
+      (* CR lmaurer: Re-enable this test if we can figure out how to square
+         the [lib-dynlink-native] test with it. *)
+      if false && not access_allowed then
+        let prefix = Compilation_unit.for_pack_prefix current_unit in
+        error (Direct_reference_from_wrong_package (name, filename, prefix));
```
[the remark this answers is not in the document]
Comment (lukemaurer) [reply ocaml/typing/persistent_env.ml:207]: I re-enabled the check having fixed `lib-dynlink-native`, as discussed yesterday.

File: ocaml/typing/persistent_env.ml [hunk 10]
```diff
@@ -193,6 +197,18 @@ let acknowledge_pers_struct penv check modname pers_sig pm =
...
+  | _ -> ()
```
[the remark this answers is not in the document]
Comment (lukemaurer) [reply ocaml/typing/persistent_env.ml:208]: Why indeed. Changed.

File: ocaml/typing/types.ml (hunk 5 above)
[the remark this answers is not in the document]
Comment (lpw25) [reply ocaml/typing/types.ml:104]: I think that it probably uses a copy of this file, so it's actually the API that it is using.

File: ocaml/typing/typemod.ml (hunk 6 above)
[the remark this answers is not in the document]
Comment (lpw25) [reply ocaml/typing/typemod.ml:1988]: I can't see this change. Missing push?

File: ocaml/typing/env.ml [hunk 11]
```diff
@@ -735,7 +751,13 @@ let find_same_module id tbl =
...
+      (* CR lmaurer: The use of [Ident.create_persistent] here is pretty gross in
+         the current regime. It's not actually the [Ident.t] corresponding to a
+         compilation unit (and only bytecode compilation guarantees there is
+         such a thing). It's really just part of the API between this and
+         [Typemod.type_module_aux]. *)
```
[the remark this answers is not in the document]
Comment (lukemaurer) [reply ocaml/typing/env.ml:760]: Deleted the comment. I think I understand the role that `Ident.t` plays a bit better now. (Also, I can't work out what I was getting at about `type_module_aux`; might have misread something.) The name `create_persistent` does seem misleading now, though: it's not actually a name that gets saved anywhere (besides in bytecode, and even then only if the named unit isn't in a pack).

**Review by mshinwell — APPROVED** [review]
This is a great change which has cleaned up a lot of code.
Approving everything except:
- the typing/ directory
- the comment on Compilation_unit.can_access_by_name (although I suggest renaming this function to make its semantics clearer)
- the add_used_primitive / get_units_with_used_primitives logic in Translprim.
@lukemaurer please liaise with @lpw25 for these final parts of review, as discussed.

**Review by lpw25 — APPROVED** [review]
Approving the parts Mark didn't
